src/persist-client/src/internal/state.rs

```diff
@@ -285,36 +293,53 @@ where
             lease_duration_ms: u64::try_from(lease_duration.as_millis())
                 .expect("lease duration as millis should fit within u64"),
         };
+
+        // If the shard-global upper and since are both the empty antichain,
+        // then no further writes can ever commit and no further reads can be
+        // served. Optimize this by no-op-ing reader registration so that we can
+        // settle the shard into a final unchanging tombstone state.
+        if self.is_tombstone() {
+            assert!(self.leased_readers.is_empty());
```

Comment:
happy to debate the merits, but fwiw these are intentional. my goals were 1) so that a code reader sees the comment and the assert and this (hopefully) allows them to reason about the correctness of the tombstone special case without having to follow down some code rabbitholes too far and 2) to prevent hypothetical future bugs when someone changes the common tombstone methods

(1) is a judgement call and you and paul are really the arbiters of whether it was effective (sounds like "no" for you). (2) is relevant here because i'm particularly stressed about the overall correctness of this tombstone stuff (it's hard to write it in a way that allows you to reason about it entirely locally and also it's definitely gonna get less test coverage than the more commonly used bits of persist)

going back and rereading the comment as the PR ended up, I find (1) not entirely obvious, so perhaps the resolution here is to tweak the comment somehow

Reply:
I guess I'd agree that it's not clear to me that "there are no readers" is the most salient info for reasoning about the correctness of the "register a new reader" method. But that likely says more about me than the code of course.

In either case, I don't think this is something that particularly needs resolving! Just getting a feel for the thinking here.

Reply:
deleted them! certainly seems the evidence is they were most confusing then enlightening in practice